### ISSUE TITLE
Showing whether a notebook is busy in the tabs icons

### ISSUE THREAD

**Issue by Maximophone** [jupyterlab/jupyterlab#4758]
In the old jupyter notebooks, when calculations were running in a cell, the browser tab's icon would change. This is super useful when you deal with long calculations all the time and have to multi-task. You do not want to switch back repeatedly to the tab where the calculation is running to check whether it is finished. Would it be easy to reimplement this feature in Jupyter Lab?

Here are screenshots of what I am talking about, in the old notebooks:

Idle notebook: ![screen shot 2018-06-21 at 13 09 51](https://user-images.githubusercontent.com/6048492/41718944-64030f24-7556-11e8-92b7-5e3b61831e3e.png)
Busy notebook: ![screen shot 2018-06-21 at 13 09 38](https://user-images.githubusercontent.com/6048492/41718946-65f35aa0-7556-11e8-8a78-939fce5bfd54.png)

**Comment by jasongrout** [comment]
This is done in #4361, and should be in the next version of JupyterLab. Thanks!

**Comment by afshin** [comment]
@jasongrout I spoke with @Maximophone about this. https://github.com/jupyterlab/jupyterlab/pull/4361 updates the browser window's favicon, but **within a JupyterLab session**, the user does not know if another JupyterLab tab is busy or idle. So in a sense, we still are not at parity with the classic notebook, because we can't tell which notebook is currently busy.

**Comment by telamonian** [comment]
What's weird is that the command bar that sits below the tab bar seems to know if a kernel is busy:

idle:
![image](https://user-images.githubusercontent.com/2263641/43745627-dd30f886-99ad-11e8-9e00-31416de05809.png)

busy:
![image](https://user-images.githubusercontent.com/2263641/43745643-eac7cc2c-99ad-11e8-89cf-28d366a3079c.png)

So I guess it would just be a matter of copying the code running the animation in the command bar over to the code running the tabs.

**Comment by telamonian** [comment]
One complication is that the tabs use a similar filled circle symbol to indicate unsaved changes in a file:
![image](https://user-images.githubusercontent.com/2263641/43745872-14000bc6-99af-11e8-97a8-4c77c603b89f.png)

For the sake of consistent symbol use/meaning (semantics, I guess?) I think one of those symbols needs to change. Maybe use a tilde for unsaved file changes instead?

**Comment by jasongrout** [comment]
Another standard thing to do is put a `*` next to the filename to indicate the file is dirty.

**Comment by orena1** [comment]
Any update on this one?

**Comment by jasongrout** [comment]
I don't think so. Any update should have been noted here. Do you want to work on it?

**Comment by orena1** [comment]
Don't know JS enough...
I'm trying jupyterlab every 6 months to see if I should move from notebook. Its getting better, but not there yet, thanks for all the work!

**Comment by jasongrout** [comment]
Note that the browser tab already indicates whether some notebook that is open in jlab is busy (so it's the same as the classic Notebook), but we don't have individual tabs inside of JupyterLab showing busy status (that's this issue).

**Comment by GabrielaVives** [comment]
Hi All,

I have worked on the UX/UI of this subject, and I agree that putting a * next to the file name is the standard when showing the user that a file has unsaved changes.

I suggest to use that, and use the dot to show the changes about the status: something is happening in the tab or not.

I illustrated all the use cases and the UI design below :

<img width="666" alt="image" src="https://user-images.githubusercontent.com/105775162/220122044-bdc117ab-a49d-4edd-9f18-f7be3233557b.png">

**Comment by maxxu05** [comment]
This would be great, if I have a lot of notebooks open, it is very difficult to see which one is actually running

**Comment by krassowski** [comment]
I would be confused by the proposed design. What if instead we had the tab background (or part of it) work as a progess bar? That would bring more value than barely running/not running. Users are also already familiar with this in context of some operating systems.

**Comment by GabrielaVives** [comment]
> This would be great, if I have a lot of notebooks open, it is very difficult to see which one is actually running. Have there been any updates on this since 2020?

Do you mean that you would have only one notebook running at a time? Are you talking in general or referring to the design suggested above?

**Comment by xgdgsc** [comment]
I would prefer changing the icon at left to the dot instead.

**Comment by maxxu05** [comment]
Apologies on late response:
@krassowski I don't know if a progress bar is possible? That implies that we would always know what the progress is while running a cell.
@GabrielaVives To clarify: I have multiple notebooks running, and I am referring in general, not necessarily to that design. Basically, in jupyter notebook, when we had multiple notebooks open in multiple tabs, we could exactly see which notebook was running a cell, because the jupyter icon would change to an hourglass for each tab that was running then revert back when done. 

However, currently, if I have multiple notebooks running in jupyter lab, none of the individual notebook's icon's (in their tab) change to indicate when they are running/done running. They only show the dot that says latest notebook is not saved yet.

Please let me know if this makes sense and I can clarify

**Comment by krassowski** [comment]
> I don't know if a progress bar is possible? That implies that we would always know what the progress is while running a cell.

Yes, it is. We know how many cells are scheduled and how many have completed. This is already used for progress bar (packman-style by default) in the busy kernel indicator. In the future we might even have a progress on per-cell basis, but as of today there is no [enhancement proposal](https://github.com/jupyter/enhancement-proposals) although something akin to this idea was discussed [back in 2018](https://github.com/jupyter/jupyter_client/issues/376).

A quick mockup with the kernel in active notebook tab (`Untitled9.ipynb`) having executed 1/2 cells and the notebook in the background (`Untitled10.ipynb`) having executed 3/4 cells scheduled:

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/be4ee006-f72b-4123-bc28-cbc1c64e664e)

The active notebook was also modified and has unsaved changes.

**Comment by krassowski** [comment]
A bit less visually appealing idea would be having packmans everywhere (e.g. replacing the notebook icon):

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/b0e48461-5c58-49a9-9288-a75647acd03a)

Maybe if it was somewhat modified to match the colours of the notebook icon it could be more palatable.

**Comment by krassowski** [comment]
> I would prefer changing the icon at left to the dot instead.

@xgdgsc can you please clarify?

**Comment by xgdgsc** [comment]
Your latest screenshot is great.

**Comment by GabrielaVives** [comment]
I'm concerned that this would  be useful only when a few tabs are open. But if there are 30 tabs open, the tab will be reduced to its minimum width and the progress bar won't signify much any more. It could bring non-necessary visual distractions: many progress bars movings along at different paces.

Would a spinner similar to version in the notebook be an option? However, it wouldn't inform on the amount of processing completed/the total time. Maybe this information could be displayed in a tooltip (maybe a percentage?) when hovering over the spinner icon? This way users could see the progress only when they need too.

**Comment by krassowski** [comment]
> I'm concerned that this would be useful only when a few tabs are open. But if there are 30 tabs open, the tab will be reduced to its minimum width and the progress bar won't signify much any more.

Not a problem once support for scrolling in the tabbar  (https://github.com/jupyterlab/jupyterlab/pull/12278) is merged.

> It could bring non-necessary visual distractions: many progress bars movings along at different paces.

Possibly, and we should certainly allow user to disable this. But then the vast majority of users, even if they have 30+ tabs open, will only have execution going in a few of them (most frequently one or two).

> Would a spinner similar to version in the notebook be an option?

There is certainly a few options along this idea to be explored (a small progress bar in the icon, a pulsating dot in the icon etc).
